### ISSUE TITLE
Bring back multiple output for binaries

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#32082]
@peti reverted https://github.com/NixOS/nixpkgs/pull/27209 since it broke haskell packages

The fix is quite simple:

```
diff --git a/pkgs/development/haskell-modules/with-packages-wrapper.nix b/pkgs/development/haskell-modules/with-packages-wrapper.nix
index ec7d961581a..725a2812f94 100644
--- a/pkgs/development/haskell-modules/with-packages-wrapper.nix
+++ b/pkgs/development/haskell-modules/with-packages-wrapper.nix
@@ -43,7 +43,7 @@ let
   libDir        = if isHaLVM then "$out/lib/HaLVM-${ghc.version}" else "$out/lib/${ghcCommand}-${ghc.version}";
   docDir        = "$out/share/doc/ghc/html";
   packageCfgDir = "${libDir}/package.conf.d";
-  paths         = lib.filter (x: x ? isHaskellLibrary) (lib.closePropagation packages);
+  paths         = map lib.getLib (lib.filter (x: x ? isHaskellLibrary) (lib.closePropagation packages));
   hasLibraries  = lib.any (x: x.isHaskellLibrary) paths;
   # CLang is needed on Darwin for -fllvm to work:
   # https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/code-generators.html
```

and a proof:

```
$ nix-shell -p "haskellPackages.ghcWithPackages (ps: [ps.mtl])" --run "ghc-pkg list mtl" -I ~/dev
these derivations will be built:
  /nix/store/ri4g0qry776qgirvw4wm4xb7yqsfbagv-ghc-8.0.2-with-packages.drv
building path(s) ‘/nix/store/s8ja7k00cd9ffd0h9prcaivmjgvdh8bd-ghc-8.0.2-with-packages’
/nix/store/s8ja7k00cd9ffd0h9prcaivmjgvdh8bd-ghc-8.0.2-with-packages/lib/ghc-8.0.2/package.conf.d
    mtl-2.2.1
```

cc @nc6

**Comment by Fuuzetsu** [comment]
It needs testing on hydra that it doesn't break anything else... I have wasted days fixing one thing and breaking another in this code ;/

**Comment by nc6** [comment]
I'll resurrect the PR, I did have a somewhat updated version

**Comment by nc6** [comment]
https://github.com/NixOS/nixpkgs/pull/32112 opened with your fix, needs hydra testing.

**Comment by vcunat** [comment]
Merged to master, but not all caveats seems solved yet.

**Comment by domenkozar** [comment]
@vcunat what's missing?

**Comment by vcunat** [comment]
@domenkozar: see the end of that PR's discussion: https://github.com/NixOS/nixpkgs/pull/32112

**Comment by domenkozar** [comment]
FYI: I have fixed ghcWithPackages and a few linux build failures.

What is left is:

- ghcjsHEAD 
- agda-prelude 
- happy on darwin
- cabal2nix postInstall fixes
- tests

**Comment by domenkozar** [comment]
https://github.com/NixOS/nixpkgs/pull/32629

**Comment by domenkozar** [comment]
https://github.com/NixOS/nixpkgs/pull/58525 has landed and I've backported it to 18.09 and 19.03.

I think enabling this by default is big effort, but it's great that this feature is now available as a better way to `justStaticExecutable`. Maybe that should be now removed or do the bin splitting?

**Comment by chreekat** [comment]
@domenkozar would it be correct to say this issue can be closed, perhaps if you open a more specific issue about any remaining quibbles?
